File: haskell/tests/Test_WeightedList.hs
module Test_WeightedList where

import Test.Tasty
import Test.Tasty.HUnit
import Test.Tasty.ExpectedFailure

import Data.Tuple

import Syntax
import WeightedList


---------------------------------------------------------------------

test_weighted_list :: TestTree
test_weighted_list = testGroup "WeightedList"
  [ test_collection "constructor" test_constructor
  , test_collection "properties" test_properties
  , test_collection "indexing" test_indexing
  , test_collection "merging" test_merging
  ]

test_weighted_list_errors :: TestTree
test_weighted_list_errors = expectFail $ testGroup "WeightedList Errors"
  [ test_collection "indexing" test_indexing_errors
  ]


---------------------------------------------------------------------

__ = newWeightedList @String @Int []

l  = [(2, "sup"), (3, "nova"), (7, "shard")]

wl :: WeightedList String Int
wl = newWeightedList l


---------------------------------------------------------------------

test_constructor :: [Assertion]
test_constructor =
  [
    __ === []

  , wl === [ WeightedItem { value = "sup", weight = 2, c_weight = 0 }
           , WeightedItem { value = "nova", weight = 3, c_weight = 0 }
           , WeightedItem { value = "shard", weight = 7, c_weight = 0 }
           ]
  ]

test_properties :: [Assertion]
test_properties =
  [
    total_values wl === 3
  , total_weights wl === 12
  , values wl === ["sup", "nova", "shard"]
  , weights wl === [2, 3, 7]
  , raw wl === l
  , raw' wl === map swap l
  ]

test_indexing :: [Assertion]
test_indexing =
  [
    value (get wl 0)  === "sup"
  , value (get wl 1)  === "sup"
  , value (get wl 2)  === "nova"
  , value (get wl 3)  === "nova"
  , value (get wl 4)  === "nova"
  , value (get wl 5)  === "shard"
  , value (get wl 6)  === "shard"
  , value (get wl 7)  === "shard"
  , value (get wl 8)  === "shard"
  , value (get wl 9)  === "shard"
  , value (get wl 10) === "shard"
  , value (get wl 11) === "shard"

  , value (get wl (-1))  === "shard"
  , value (get wl (-2))  === "shard"
  , value (get wl (-3))  === "shard"
  , value (get wl (-4))  === "shard"
  , value (get wl (-5))  === "shard"
  , value (get wl (-6))  === "shard"
  , value (get wl (-7))  === "shard"
  , value (get wl (-8))  === "nova"
  , value (get wl (-9))  === "nova"
  , value (get wl (-10)) === "nova"
  , value (get wl (-11)) === "sup"
  , value (get wl (-12)) === "sup"
  ]

test_indexing_errors :: [Assertion]
test_indexing_errors =
  [ Just (get wl (12)) === Nothing
  , Just (get wl (-13)) === Nothing
  ]

test_merging :: [Assertion]
test_merging =
  [
    merge __ __ === []
  , merge wl __ === wl
  , merge __ wl === wl

    -- merge 1
  , merge wl (newWeightedList [ (1, "sup") ])
          === newWeightedList [ (3, "sup"), (3, "nova"), (7, "shard") ]

    -- merge 3
  , merge wl wl === newWeightedList [ (4, "sup"), (6, "nova"), (14, "shard") ]
  
    -- append 1
  , merge wl (newWeightedList [ (13, "cortex") ])
          === newWeightedList [ (2, "sup"), (3, "nova"), (7, "shard"), (13, "cortex") ]
  
    -- append 2
  , merge wl (newWeightedList [ (13, "cortex"), (20, "origin") ])
          === wl ++ newWeightedList [ (13, "cortex"), (20, "origin") ]
  
    -- append 3
  , merge wl (newWeightedList [ (13, "cortex"), (20, "origin"), (42, "vision") ])
          === wl ++ newWeightedList [ (13, "cortex"), (20, "origin"), (42, "vision") ]
  ]

File: rust/tests/test-weighted-list.rs
use weighted_list::WeightedList;


fn data_string() -> [(i32, String); 3]
{
    [
        (2, String::from("sup")),
        (3, String::from("nova")),
        (5, String::from("shard")),
    ]
}

fn el() -> WeightedList<String, i32>
{
    WeightedList::empty()
}

fn wl() -> WeightedList<String, i32>
{
    WeightedList::from(data_string())
}


#[test]
fn testtconstructors()
{
    let _: WeightedList<String, i32> = el();

    let _: WeightedList<String, i32> = wl();

    let _: WeightedList<bool, f64> = WeightedList::from([
        (2.0, false),
        (4.2, true),
    ]);
}

#[test]
fn testtequality()
{
    assert!( el() == el() );
    assert!( wl() == wl() );

    assert!( el() != wl() );
    assert!( wl() != el() );
}

#[test]
fn testtproperties()
{
    assert!( el().len() == 0 );
    assert!( wl().len() == 10 );

    let e: [&String; 0] = [];
    assert!( el().values().eq(e) );
    assert!( wl().values().eq(["sup", "nova", "shard"]) );

    let e: [i32; 0] = [];
    assert!( el().weights().eq(e) );
    assert!( wl().weights().eq([2, 3, 5]) );

    let e: [(i32, &String); 0] = [];
    assert!( el().raw().eq(e) );
    assert!( wl().raw().eq([
        (2, &"sup".to_string()),
        (3, &"nova".to_string()),
        (5, &"shard".to_string()),
    ]) );
}
